book/src/foreign-function-interface.md

```diff
@@ -167,254 +165,122 @@ Steps
+(The LLVM IR can be generated using `cargo rustc --bin qoi-bindgen --release -- --emit=llvm-ir`)
+
+The solution is to use `std::mem::MaybeUninit`:
 
-%_3.i = call i32 @bindings::crypto_hash_sha512_tweet(i8* nonnull %0, i8* nonnull "foobarbaz", i64 9), !noalias !6
+```rust
+let mut desc = MaybeUninit::uninit();
+let pointer = unsafe { qoi_read(filename.as_ptr(), desc.as_mut_ptr(), 4) };
+let desc = unsafe { desc.assume_init() };
 ```
-<details>
-<summary>Full LLVM IR</summary>
-<p>
 
-```llvm
-define i8 @call_with_maybeuninit() unnamed_addr #1 personality i32 (i32, i32, i64, %"unwind::libunwind::_Unwind_Exception"*, %"unwind::libunwind::_Unwind_Context"*)* @rust_eh_personality {
-start:
-  %result.i = alloca <64 x i8>, align 1
-  %0 = getelementptr inbounds <64 x i8>, <64 x i8>* %result.i, i64 0, i64 0
-  call void @llvm.lifetime.start.p0i8(i64 64, i8* nonnull %0), !noalias !2
-  %_3.i = call i32 @crypto_hash_sha512_tweet(i8* nonnull %0, i8* nonnull getelementptr inbounds (<{ [9 x i8] }>, <{ [9 x i8] }>* @alloc1, i64 0, i32 0, i64 0), i64 9), !noalias !6
-  %1 = load <64 x i8>, <64 x i8>* %result.i, align 1, !noalias !7
-  call void @llvm.lifetime.end.p0i8(i64 64, i8* nonnull %0), !noalias !2
-  %2 = call i8 @llvm.vector.reduce.add.v64i8(<64 x i8> %1)
-  ret i8 %2
-}
+The `MaybeUninit` type is an abstraction for uninitialized memory. The `.uninit()` method gives a chunk of uninitialized memory big enough to store a value of the desired type (in our case `qoi_desc` will be inferred).
 
-define i8 @call_without_maybeuninit() unnamed_addr #1 personality i32 (i32, i32, i64, %"unwind::libunwind::_Unwind_Exception"*, %"unwind::libunwind::_Unwind_Context"*)* @rust_eh_personality {
-start:
-  %_4 = alloca <64 x i8>, align 1
-  %0 = getelementptr inbounds <64 x i8>, <64 x i8>* %_4, i64 0, i64 0
-  call void @llvm.lifetime.start.p0i8(i64 64, i8* nonnull %0)
-  call void @llvm.memset.p0i8.i64(i8* noundef nonnull align 1 dereferenceable(64) %0, i8 0, i64 64, i1 false), !alias.scope !8, !noalias !11
-  %_2.i = call i32 @crypto_hash_sha512_tweet(i8* nonnull %0, i8* nonnull getelementptr inbounds (<{ [9 x i8] }>, <{ [9 x i8] }>* @alloc1, i64 0, i32 0, i64 0), i64 9)
-  %1 = load <64 x i8>, <64 x i8>* %_4, align 1
-  %2 = call i8 @llvm.vector.reduce.add.v64i8(<64 x i8> %1)
-  call void @llvm.lifetime.end.p0i8(i64 64, i8* nonnull %0)
-  ret i8 %2
-}
-```
+### Conclusion
+In this exercise we saw how we can generate bindings to a C library with bindgen. The generated bindings are a bit difficult to work with, as they are unsafe and rely on C types. We've discussed how we can create nice wrappers around the generated bindings to deal with all these C types and to make them safer to work with.
 
-</p>
-</details>
+We should also note that the image crate has native support for QOI images, so in reality reading a QOI file could be as simple as `image::open("image.qoi")` without needing to load an external C library 😉.
```

Comment:
"We should also note that the image crate ...  " maybe link to the crate / docs.rs page? 

Also, instead of just saying "hey we just wasted your time", we could just be open about this up front and add an actual test case that confirms that the image loaded with `qoi.h` is the same as the one that the image crate reads?

Reply:
Good suggestion! I've added a test and I've rewritten the text to be up front about this.